README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Memcached components for Pip.Services in Node.js / ES2017

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Memcached module contains the following components: MemcachedLock and MemcachedCache for working with locks and cache on the Memcached server.

The module contains the following packages:
- **Build** - a standard factory for constructing components.
- **Cache** - cache Components in Memcached
- **Lock** - components of working with locks in Memcached

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pip-services3-nodex.github.io/pip-services3-memcached-nodex/globals.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Install the NPM package as
```bash
npm install pip-services-memcached-nodex --save
```

## Develop

For development you shall install the following prerequisites:
* Node.js 14+
* Visual Studio Code or another IDE of your choice
* Docker
* Typescript

Install dependencies:
```bash
npm install
```

Compile the code:
```bash
tsc
```

Run automated tests:
```bash
npm test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov**.


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Memcached components for Pip.Services in Node.js / ES2017 Changelog

## <a name="1.0.1"></a> 1.0.1 (2021-10-29) 
### Bug fixing
* Fix working with objects, added tests
## <a name="1.0.0"></a> 1.0.0 (2021-05-28) 

### Features
* **memcached** Added MemcachedCache
* **memcached** Added MemcachedLock



test/fixtures/LockFixture.ts
const assert = require('chai').assert;

import { ILock } from 'pip-services3-components-nodex';

let LOCK1: string = "lock_1";
let LOCK2: string = "lock_2";
let LOCK3: string = "lock_3";

export class LockFixture {
    private _lock: ILock;

    public constructor(lock: ILock) {
        this._lock = lock;
    }

    public async testTryAcquireLock(): Promise<void> {
        // Try to acquire lock for the first time
        let ok = await this._lock.tryAcquireLock(null, LOCK1, 3000);
        assert.isTrue(ok);

        // Try to acquire lock for the second time
        ok = await this._lock.tryAcquireLock(null, LOCK1, 3000);
        assert.isFalse(ok);

        // Release the lock
        await this._lock.releaseLock(null, LOCK1);

        // Try to acquire lock for the third time
        ok = await this._lock.tryAcquireLock(null, LOCK1, 3000);
        assert.isTrue(ok);

        await this._lock.releaseLock(null, LOCK1);
    }

    public async testAcquireLock(): Promise<void> {
        // Acquire lock for the first time
        await this._lock.acquireLock(null, LOCK2, 3000, 1000);

        // Acquire lock for the second time
        try {
            await this._lock.acquireLock(null, LOCK2, 3000, 1000);
            assert.fail("Expected exception on the second lock attempt");
        } catch {
            // Expected exception...
        }

        // Release the lock
        await this._lock.releaseLock(null, LOCK2);

        // Acquire lock for the third time
        await this._lock.acquireLock(null, LOCK2, 3000, 1000);

        await this._lock.releaseLock(null, LOCK2);
    }

    public async testReleaseLock(): Promise<void> {
        // Acquire lock for the first time
        let ok = await this._lock.tryAcquireLock(null, LOCK3, 3000);
        assert.isTrue(ok);

        // Release the lock for the first time
        await this._lock.releaseLock(null, LOCK3);

        // Release the lock for the second time
        await this._lock.releaseLock(null, LOCK3);
    }
    
}

test/fixtures/CacheFixture.ts
const assert = require('chai').assert;

import { ICache } from 'pip-services3-components-nodex';

let KEY1: string = "key1";
let KEY2: string = "key2";
let KEY3: string = "key3";
let KEY4: string = "key4";
let KEY5: string = "key5";
let KEY6: string = "key6";

let VALUE1: string = "value1";
let VALUE2 = { val: "value2" };
let VALUE3 = new Date();
let VALUE4 = [1, 2, 3, 4];
let VALUE5 = 12345;
let VALUE6 = null;


export class CacheFixture {
    private _cache: ICache = null;

    public constructor(cache: ICache) {
        this._cache = cache;
    }

    public async testStoreAndRetrieve(): Promise<void> {
        await this._cache.store(null, KEY1, VALUE1, 5000);
        await this._cache.store(null, KEY2, VALUE2, 5000);
        await this._cache.store(null, KEY3, VALUE3, 5000);
        await this._cache.store(null, KEY4, VALUE4, 5000);
        await this._cache.store(null, KEY5, VALUE5, 5000);
        await this._cache.store(null, KEY6, VALUE6, 5000);

        await new Promise(resolve => setTimeout(resolve, 500));

        let val = await this._cache.retrieve(null, KEY1)
        assert.isNotNull(val);
        assert.equal(VALUE1, val);

        val = await this._cache.retrieve(null, KEY2);
        assert.isNotNull(val);
        assert.equal(VALUE2.val, val.val);

        val = await this._cache.retrieve(null, KEY3);
        assert.isNotNull(val);
        assert.equal(VALUE3.toISOString(), val);

        val = await this._cache.retrieve(null, KEY4);
        assert.isNotNull(val);
        assert.lengthOf(val, 4)
        assert.equal(VALUE4[0], val[0]);

        val = await this._cache.retrieve(null, KEY5)
        assert.isNotNull(val);
        assert.equal(VALUE5, val);

        val = await this._cache.retrieve(null, KEY6);
        assert.isNull(val);
    }

    public async testRetrieveExpired(): Promise<void> {
        this._cache.store(null, KEY1, VALUE1, 1000);

        await new Promise(resolve => setTimeout(resolve, 1500));

        let val = await this._cache.retrieve(null, KEY1);
        assert.isNull(val || null);
    }
    
    public async testRemove(): Promise<void> {
        await this._cache.store(null, KEY1, VALUE1, 1000);

        await this._cache.remove(null, KEY1);

        let val = await this._cache.retrieve(null, KEY1);
        assert.isNull(val || null);
    }

}


test/lock/MemcachedLock.test.ts
import { ConfigParams } from 'pip-services3-commons-nodex';

import { MemcachedLock } from '../../src/lock/MemcachedLock';
import { LockFixture } from '../fixtures/LockFixture';

suite('MemcachedLock', ()=> {
    let _lock: MemcachedLock;
    let _fixture: LockFixture;

    setup(async () => {
        let host = process.env['MEMCACHED_SERVICE_HOST'] || 'localhost';
        let port = process.env['MEMCACHED_SERVICE_PORT'] || 11211;

        _lock = new MemcachedLock();

        let config = ConfigParams.fromTuples(
            'connection.host', host,
            'connection.port', port
        );
        _lock.configure(config);

        _fixture = new LockFixture(_lock);

        await _lock.open(null);
    });

    teardown(async () => {
        await _lock.close(null);
    });

    test('Try Acquire Lock', async () => {
        await _fixture.testTryAcquireLock();
    });

    test('Acquire Lock', async () => {
        await _fixture.testAcquireLock();
    });

    test('Release Lock', async () => {
        await _fixture.testReleaseLock();
    });

});


src/cache/MemcachedCache.ts
/** @module cache */
import { ConfigParams } from 'pip-services3-commons-nodex';
import { IConfigurable } from 'pip-services3-commons-nodex';
import { IReferences } from 'pip-services3-commons-nodex';
import { IReferenceable } from 'pip-services3-commons-nodex';
import { IOpenable } from 'pip-services3-commons-nodex';
import { InvalidStateException } from 'pip-services3-commons-nodex';
import { ConfigException } from 'pip-services3-commons-nodex';
import { ConnectionResolver } from 'pip-services3-components-nodex';
import { ICache } from 'pip-services3-components-nodex';

/**
 * Distributed cache that stores values in Memcaches caching service.
 * 
 * The current implementation does not support authentication.
 * 
 * ### Configuration parameters ###
 * 
 * - connection(s):           
 *   - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]]
 *   - host:                  host name or IP address
 *   - port:                  port number
 *   - uri:                   resource URI or connection string with all parameters in it
 * - options:
 *   - max_size:              maximum number of values stored in this cache (default: 1000)        
 *   - max_key_size:          maximum key length (default: 250)
 *   - max_expiration:        maximum expiration duration in milliseconds (default: 2592000)
 *   - max_value:             maximum value length (default: 1048576)
 *   - pool_size:             pool size (default: 5)
 *   - reconnect:             reconnection timeout in milliseconds (default: 10 sec)
 *   - retries:               number of retries (default: 3)
 *   - timeout:               default caching timeout in milliseconds (default: 1 minute)
 *   - failures:              number of failures before stop retrying (default: 5)
 *   - retry:                 retry timeout in milliseconds (default: 30 sec)
 *   - idle:                  idle timeout before disconnect in milliseconds (default: 5 sec)
 * 
 * ### References ###
 * 
 * - <code>\*:discovery:\*:\*:1.0</code>        (optional) [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]] services to resolve connection
 *
 * ### Example ###
 * 
 *     let cache = new MemcachedCache();
 *     cache.configure(ConfigParams.fromTuples(
 *       "host", "localhost",
 *       "port", 11211
 *     ));
 * 
 *     await cache.open("123");
 * 
 *     await cache.store("123", "key1", "ABC");
 *     let value = await cache.store("123", "key1"); // Result: "ABC"
 */
export class MemcachedCache implements ICache, IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();

    private _maxKeySize: number = 250;
    private _maxExpiration: number = 2592000;
    private _maxValue: number = 1048576;
    private _poolSize: number = 5;
    private _reconnect: number = 10000;
    private _timeout: number = 5000;
    private _retries: number = 5;
    private _failures: number = 5;
    private _retry: number = 30000;
    private _remove: boolean = false;
    private _idle: number = 5000;

    private _client: any = null;

    /**
     * Creates a new instance of this cache.
     */
    public constructor() { }

    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        this._connectionResolver.configure(config);

        this._maxKeySize = config.getAsIntegerWithDefault('options.max_key_size', this._maxKeySize);
        this._maxExpiration = config.getAsLongWithDefault('options.max_expiration', this._maxExpiration);
        this._maxValue = config.getAsLongWithDefault('options.max_value', this._maxValue);
        this._poolSize = config.getAsIntegerWithDefault('options.pool_size', this._poolSize);
        this._reconnect = config.getAsIntegerWithDefault('options.reconnect', this._reconnect);
        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
        this._failures = config.getAsIntegerWithDefault('options.failures', this._failures);
        this._retry = config.getAsIntegerWithDefault('options.retry', this._retry);
        this._remove = config.getAsBooleanWithDefault('options.remove', this._remove);
        this._idle = config.getAsIntegerWithDefault('options.idle', this._idle);
    }

    /**
     * Sets references to dependent components.
     * 
     * @param references 	references to locate the component dependencies. 
     */
    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
    }

    /**
     * Checks if the component is opened.
     * 
     * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._client;
    }

    /**
     * Opens the component.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async open(correlationId: string): Promise<void> {
        let connections = await this._connectionResolver.resolveAll(correlationId);
        if (connections.length == 0) {
            throw new ConfigException(
                correlationId,
                'NO_CONNECTION',
                'Connection is not configured'
            );
        }

        let servers: string[] = [];
        for (let connection of connections) {
            let host = connection.getHost();
            let port = connection.getPort() || 11211;
            servers.push(host + ':' + port);
        }

        let options = {
            maxKeySize: this._maxKeySize,
            maxExpiration: this._maxExpiration,
            maxValue: this._maxValue,
            poolSize: this._poolSize,
            reconnect: this._reconnect,
            timeout: this._timeout,
            retries: this._retries,
            failures: this._failures,
            retry: this._retry,
            remove: this._remove,
            idle: this._idle
        };

        let Memcached = require('memcached');
        this._client = new Memcached(servers, options);
    }

    /**
     * Closes component and frees used resources.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async close(correlationId: string): Promise<void> {
        this._client = null;
    }

    private checkOpened(correlationId: string): void {
        if (!this.isOpen()) {
            throw new InvalidStateException(
                correlationId,
                'NOT_OPENED',
                'Connection is not opened'
            );
        }
    }

    /**
     * Retrieves cached value from the cache using its key.
     * If value is missing in the cache or expired it returns null.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @returns a cached value or <code>null</code> if nothing was found.
     */
    public retrieve(correlationId: string, key: string): Promise<any> {
        this.checkOpened(correlationId);

        return new Promise<any>((resolve, reject) => {
            this._client.get(key, (err, result) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve(result ? JSON.parse(result) : result);
            });
        });
    }

    /**
     * Stores value in the cache with expiration time.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @param value             a value to store.
     * @param timeout           expiration timeout in milliseconds.
     * @returns the stored value
     */
    public store(correlationId: string, key: string, value: any, timeout: number): Promise<any> {
        this.checkOpened(correlationId);

        let timeoutInSec = timeout / 1000;

        return new Promise<any>((resolve, reject) => {
            this._client.set(key, JSON.stringify(value), timeoutInSec, (err, result) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve(result);
            });
        });
    }

    /**
     * Removes a value from the cache by its key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @returns the deleted value.
     */
    public remove(correlationId: string, key: string): Promise<any> {
        this.checkOpened(correlationId);

        return new Promise<any>((resolve, reject) => {
            this._client.del(key, (err, result) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve(result ? JSON.parse(result) : result);
            });
        });
    }

}

src/lock/MemcachedLock.ts
/** @module lock */
import { ConfigParams } from 'pip-services3-commons-nodex';
import { IConfigurable } from 'pip-services3-commons-nodex';
import { IReferences } from 'pip-services3-commons-nodex';
import { IReferenceable } from 'pip-services3-commons-nodex';
import { IOpenable } from 'pip-services3-commons-nodex';
import { InvalidStateException } from 'pip-services3-commons-nodex';
import { ConfigException } from 'pip-services3-commons-nodex';
import { ConnectionResolver } from 'pip-services3-components-nodex';
import { Lock } from 'pip-services3-components-nodex';

/**
 * Distributed lock that implemented based on Memcaches caching service.
 * 
 * The current implementation does not support authentication.
 * 
 * ### Configuration parameters ###
 * 
 * - connection(s):           
 *   - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]]
 *   - host:                  host name or IP address
 *   - port:                  port number
 *   - uri:                   resource URI or connection string with all parameters in it
 * - options:
 *   - retry_timeout:         timeout in milliseconds to retry lock acquisition. (Default: 100)
 *   - max_size:              maximum number of values stored in this cache (default: 1000)        
 *   - max_key_size:          maximum key length (default: 250)
 *   - max_expiration:        maximum expiration duration in milliseconds (default: 2592000)
 *   - max_value:             maximum value length (default: 1048576)
 *   - pool_size:             pool size (default: 5)
 *   - reconnect:             reconnection timeout in milliseconds (default: 10 sec)
 *   - retries:               number of retries (default: 3)
 *   - timeout:               default caching timeout in milliseconds (default: 1 minute)
 *   - failures:              number of failures before stop retrying (default: 5)
 *   - retry:                 retry timeout in milliseconds (default: 30 sec)
 *   - idle:                  idle timeout before disconnect in milliseconds (default: 5 sec)
 * 
 * ### References ###
 * 
 * - <code>\*:discovery:\*:\*:1.0</code>        (optional) [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]] services to resolve connection
 *
 * ### Example ###
 * 
 *     let lock = new MemcachedLock();
 *     lock.configure(ConfigParams.fromTuples(
 *       "host", "localhost",
 *       "port", 11211
 *     ));
 * 
 *     await lock.open("123");
 * 
 *     await lock.acquire("123", "key1");
 *     try {
 *       // Processing...
 *     } finally {
 *       await lock.releaseLock("123", "key1");
 *     }
 */
export class MemcachedLock extends Lock implements IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();
    
    private _maxKeySize: number = 250;
    private _maxExpiration: number = 2592000;
    private _maxValue: number = 1048576;
    private _poolSize: number = 5;
    private _reconnect: number = 10000;
    private _timeout: number = 5000;
    private _retries: number = 5;
    private _failures: number = 5;
    private _retry: number = 30000;
    private _remove: boolean = false;
    private _idle: number = 5000;

    private _client: any = null;

    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        super.configure(config);

        this._connectionResolver.configure(config);

        this._maxKeySize = config.getAsIntegerWithDefault('options.max_key_size', this._maxKeySize);
        this._maxExpiration = config.getAsLongWithDefault('options.max_expiration', this._maxExpiration);
        this._maxValue = config.getAsLongWithDefault('options.max_value', this._maxValue);
        this._poolSize = config.getAsIntegerWithDefault('options.pool_size', this._poolSize);
        this._reconnect = config.getAsIntegerWithDefault('options.reconnect', this._reconnect);
        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
        this._failures = config.getAsIntegerWithDefault('options.failures', this._failures);
        this._retry = config.getAsIntegerWithDefault('options.retry', this._retry);
        this._remove = config.getAsBooleanWithDefault('options.remove', this._remove);
        this._idle = config.getAsIntegerWithDefault('options.idle', this._idle);
    }

    /**
	 * Sets references to dependent components.
	 * 
	 * @param references 	references to locate the component dependencies. 
     */
    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
    }

    /**
	 * Checks if the component is opened.
	 * 
	 * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._client;
    }

    /**
	 * Opens the component.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async open(correlationId: string): Promise<void> {
        let connections = await this._connectionResolver.resolveAll(correlationId);
        if (connections.length == 0) {
            throw new ConfigException(
                correlationId,
                'NO_CONNECTION',
                'Connection is not configured'
            );
        }

        let servers: string[] = [];
        for (let connection of connections) {
            let host = connection.getHost();
            let port = connection.getPort() || 11211;
            servers.push(host + ':' + port);
        }

        let options = {
            maxKeySize: this._maxKeySize,
            maxExpiration: this._maxExpiration,
            maxValue: this._maxValue,
            poolSize: this._poolSize,
            reconnect: this._reconnect,
            timeout: this._timeout,
            retries: this._retries,
            failures: this._failures,
            retry: this._retry,
            remove: this._remove,
            idle: this._idle
        };

        let Memcached = require('memcached');
        this._client = new Memcached(servers, options);
    }

    /**
	 * Closes component and frees used resources.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async close(correlationId: string): Promise<void> {
        this._client = null;
    }

    private checkOpened(correlationId: string): void {
        if (!this.isOpen()) {
            throw new InvalidStateException(
                correlationId,
                'NOT_OPENED',
                'Connection is not opened'
            );
        }
    }
    
    /**
     * Makes a single attempt to acquire a lock by its key.
     * It returns immediately a positive or negative result.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique lock key to acquire.
     * @param ttl               a lock timeout (time to live) in milliseconds.
     * @returns <code>true</code> if lock was successfull and <code>false</code> otherwise.
     */
    public tryAcquireLock(correlationId: string, key: string, ttl: number): Promise<boolean> {
        this.checkOpened(correlationId);

        let lifetimeInSec = ttl / 1000;
        return new Promise<boolean>((resolve, reject) => {
            this._client.add(key, 'lock', lifetimeInSec, (err) => {
                if (err != null && err.message && err.message.indexOf('not stored') >= 0) {
                    resolve(false);
                } else if (err != null) {
                    reject(err);
                } else {
                    resolve(true);
                }
            });    
        });
    }

    /**
     * Releases prevously acquired lock by its key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique lock key to release.
     */
    public releaseLock(correlationId: string, key: string): Promise<void> {
        this.checkOpened(correlationId);

        return new Promise<void>((resolve, reject) => {
            this._client.del(key, (err) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve();
            });
        });
    }
}